### PR TITLE
[Breaking] MOS: 65816 assembler fixes, part 2

### PR DESCRIPTION
**Pull request by asiekierka**
More fixes for https://github.com/llvm-mos/llvm-mos/issues/454 . These come after https://github.com/llvm-mos/llvm-mos/pull/455 .

- The immediate and address modifiers (#<, #>, #^, <, !, >) now adhere to WDC recommendations as documented in the 65816 datasheet.
  - In particular, this means the behaviour of "lda >addr" changes, while "lda !addr" and "lda #^addr" have been introduced.
- Immediate expressions now emit R_MOS_IMM8 and R_MOS_IMM16, while address expressions now emit R_MOS_ADDR8 and R_MOS_ADDR16 relocations.
- Some error messages have been adjusted to more accurately represent the assembler's limitations.
- Bugs/limitations in the SPC700 assembler's handling of "addr + X" and "addr + Y" addressing mode syntaxes have been fixed, and a test has been added.
- "S" and "R" can now be used as register names in stack-relative addressing modes instead of "SP" and "RP". This matches WDC 65816 and 65EL02 documentation/recommendations, respectively.
- Extensive tests have been added to test for addressing mode and modifier handling.

Note that this includes breaking changes for the following expressions:

- "lda >addr" was previously interpreted as "lda mos16hi(addr)", but now will be interpreted as "lda mos24(addr)". This is unlikely to break production code.
- "lda <addr8" was previously interpreted as "lda mos16lo(addr)", but now will be interpreted as "lda mos8(addr)". As the two relocations are functionally (but not semantically) equivalent, this should not break production code.
- Likewise, functional (but not semantic) changes involve the relocations emitted by immediate expressions ("lda #val").

### CODE REVIEW

File: llvm/test/MC/MOS/addressing-modes-65816.s [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    ; TODO: Immediate (16-bit)
+
+    ; Absolute
+
+    lda !addr8                  ; CHECK: ad 00 00
+                                ; CHECK: R_MOS_ADDR16		.text+0x1
+    lda !addr16                 ; CHECK: ad 00 00
+                                ; CHECK: R_MOS_ADDR16		.text+0x202
+    ; TODO: lda addr16
```
Comment (asiekierka) [inline llvm/test/MC/MOS/addressing-modes-65816.s:58]: We currently lack a 16-bit equivalent to `.zeropage`. I've left these for a later PR.

File: llvm/test/MC/MOS/addressing-modes-65816.s [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    lda #>addr24                ; CHECK: a9 00
+                                ; CHECK: R_MOS_ADDR16_HI		.text+0x30303
+    lda #^addr8                 ; CHECK: a9 00
+                                ; CHECK: R_MOS_ADDR24_BANK		.text+0x1
+    lda #^addr16                ; CHECK: a9 00
+                                ; CHECK: R_MOS_ADDR24_BANK		.text+0x202
+    lda #^addr24                ; CHECK: a9 00
+                                ; CHECK: R_MOS_ADDR24_BANK		.text+0x30303
+
+    ; TODO: Immediate (16-bit)
```
Comment (asiekierka) [inline llvm/test/MC/MOS/addressing-modes-65816.s:50]: We currently lack a way to appropriately tell the assembler to force 16-bit immediates, without using `mos16()`. I've left these for a later PR.

**Comment by mysterymath** [comment]
After merging the previous change, it looks like this branch has conflicts now and needs to be synced.

**Comment by asiekierka** [comment]
Ah, sorry; should be fixed now.

**Review by mysterymath — APPROVED** [review]

**Pull request merged**